dns/dns.go

```diff
@@ -0,0 +1,124 @@
+package dns
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/giantswarm/exporterkit"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	Namespace = "dns"
+)
+
+type Config struct {
+	Logger micrologger.Logger
+
+	Hosts []string
+}
+
+type DNSCollector struct {
+	logger micrologger.Logger
+
+	hosts []string
+
+	count      map[string]float64
+	errorCount map[string]float64
+
+	countMutex      sync.Mutex
+	errorCountMutex sync.Mutex
+
+	total      *prometheus.Desc
+	errorTotal *prometheus.Desc
+	latency    *prometheus.Desc
+}
+
+func NewCollector(config Config) (*DNSCollector, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.Hosts == nil {
+		return nil, microerror.Maskf(exporterkit.InvalidConfigError, "%T.Host must not be empty", config)
+	}
+
+	dnsCollector := DNSCollector{
```

Comment:
It would be nice to simply initiate the pointer here and not indirect below. When you need to maintain this code tomorrow and have to work with methods of the collector you have to change this and this is just more work so we could simply do it "right" straight away. I think most of our other constructors work the same way.